subsys/bluetooth/mesh/provisioner.c

```diff
@@ -0,0 +1,761 @@
+#include <sys/util.h>
+#include <sys/byteorder.h>
+
+#include <net/buf.h>
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/mesh.h>
+#include <bluetooth/uuid.h>
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_MESH_DEBUG_PROVISIONER)
+#define LOG_MODULE_NAME bt_mesh_provisioner
+#include "common/log.h"
+
+#include "host/ecc.h"
+#include "host/testing.h"
+
+#include "crypto.h"
+#include "adv.h"
+#include "mesh.h"
+#include "net.h"
+#include "rpl.h"
+#include "beacon.h"
+#include "access.h"
+#include "foundation.h"
+#include "proxy.h"
+#include "prov_common.h"
+#include "settings.h"
+#include "prov_bearer.h"
+
+struct provisioner_link {
+	struct bt_mesh_cdb_node *node;
+	uint16_t addr;
+	uint16_t net_idx;
+	uint8_t attention_duration;
+	uint8_t uuid[16];
+} provisioner;
+
+static void send_pub_key(void);
+static void prov_dh_key_gen(void);
+static void pub_key_ready(const uint8_t *pkey);
+
+static int reset_state(void)
+{
+	if (provisioner.node != NULL) {
+		bt_mesh_cdb_node_del(provisioner.node, false);
+	}
+
+	return bt_mesh_prov_reset_state(pub_key_ready);
+}
+
+static inline int prov_send(struct net_buf_simple *buf,
+			    prov_bearer_send_complete_t cb)
+{
+	return link.bearer->send(buf, cb, NULL);
+}
+
+static void prov_buf_init(struct net_buf_simple *buf, uint8_t type)
+{
+	net_buf_simple_reserve(buf, PROV_BEARER_BUF_HEADROOM);
+	net_buf_simple_add_u8(buf, type);
+}
+
+static void prov_link_close(enum prov_bearer_link_status status)
+{
+	BT_DBG("%u", status);
+	link.expect = PROV_NO_PDU;
+
+	link.bearer->link_close(status);
+}
+
+static void prov_fail(uint8_t reason)
+{
+	/* According to Bluetooth Mesh Specification v1.0.1, Section 5.4.4, the
+	 * provisioner just closes the link when something fails, while the
+	 * provisionee sends the fail message, and waits for the provisioner to
+	 * close the link.
+	 */
+	prov_link_close(PROV_BEARER_LINK_STATUS_FAIL);
+}
+
+static void send_invite(void)
+{
+	PROV_BUF(inv, 2);
+
+	BT_DBG("");
+
+	prov_buf_init(&inv, PROV_INVITE);
+	net_buf_simple_add_u8(&inv, provisioner.attention_duration);
+
+	link.conf_inputs[0] = provisioner.attention_duration;
+
+	if (prov_send(&inv, NULL)) {
+		BT_ERR("Failed to send invite");
+		return;
+	}
+
+	link.expect = PROV_CAPABILITIES;
+}
+
+static void start_sent(int err, void *cb_data)
+{
+	if (!bt_pub_key_get()) {
+		atomic_set_bit(link.flags, WAIT_PUB_KEY);
+		BT_WARN("Waiting for local public key");
+	} else {
+		send_pub_key();
+	}
+}
+
+int bt_mesh_prov_auth_start(bt_mesh_auth_method_t method,
+			    const bt_mesh_action_t *action, uint8_t size)
+{
+	BT_DBG("");
+
+	PROV_BUF(start, 6);
+	uint8_t local_method;
+	uint8_t local_action, remote_action;
+	const uint8_t *data = &link.conf_inputs[1];
+
+	prov_buf_init(&start, PROV_START);
+	net_buf_simple_add_u8(&start, PROV_ALG_P256);
+
+	if (data[3] == PUB_KEY_OOB) {
+		net_buf_simple_add_u8(&start, PUB_KEY_OOB);
+	} else {
+		net_buf_simple_add_u8(&start, PUB_KEY_NO_OOB);
+	}
```

Comment:
This should be determined by whether the application has called `bt_mesh_set_oob_public_key`. Ie, when we send start, check `REMOTE_PUB_KEY`, and request a public key if we haven't gotten one OOB. Just because a device has an OOB public key available, doesn't mean the provisioner has to follow the oob pubkey procedure.